src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -1234,10 +1236,121 @@ TargetListOnPartitionColumn(Query *query, List *targetEntryList)
 		}
 	}
 
+	/*
+	 * We could stil behave as if the target list is on partition column if
+	 * all range table entries are reference tables and there is no column
+	 * references to any outer queries.
+	 */
+	if (!targetListOnPartitionColumn)
+	{
+		if (RangeTableListContainsOnlyReferenceTables(query->rtable) &&
+			!TargetListContainsNotSupportedEntries(targetEntryList))
+		{
+			targetListOnPartitionColumn = true;
+		}
+	}
+
 	return targetListOnPartitionColumn;
 }
 
 
+/*
+ * TargetListContainsNotSupportedEntries returns true if non of the
+ * elements in the target entry list belong to an outer query (for
+ * example the query is a sublink and references to another query
+ * in the from list).
+ *
+ * The function also returns true if any of the  target entries is not
+ * a column itself. This might be too restrictive, but, given that we're
+ * handling a very specific type of queries, that seems acceptable for now.
+ */
+static bool
+TargetListContainsNotSupportedEntries(List *targetEntryList)
+{
+	ListCell *targetEntryCell = NULL;
+
+	foreach(targetEntryCell, targetEntryList)
+	{
+		TargetEntry *targetEntry = lfirst(targetEntryCell);
+		Var *candidateColumn = NULL;
+
+		Expr *strippedColumnExpression = (Expr *) strip_implicit_coercions(
+			(Node *) targetEntry->expr);
+
+		if (IsA(strippedColumnExpression, Var))
+		{
+			candidateColumn = (Var *) strippedColumnExpression;
+		}
+		else if (IsA(strippedColumnExpression, FieldSelect))
+		{
+			FieldSelect *compositeField = (FieldSelect *) strippedColumnExpression;
+			Expr *fieldExpression = compositeField->arg;
+
+			if (IsA(fieldExpression, Var))
+			{
+				candidateColumn = (Var *) fieldExpression;
+			}
+		}
+
+		/* we don't support target entries that are not columns */
+		if (candidateColumn == NULL)
+		{
+			return true;
+		}
+
+		if (candidateColumn->varlevelsup > 0)
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * RangeTableListContainsOnlyReferenceTables returns true if all relation range
+ * tables entries are reference tables.
+ *
+ * The function returns false for range table entries that are not relations.
+ *
+ * Note that the function doesn't recurse into subqueries, returns false when
+ * a subquery is found.
+ */
+static bool
+RangeTableListContainsOnlyReferenceTables(List *rangeTableList)
+{
+	ListCell *rangeTableCell = NULL;
+
+	foreach(rangeTableCell, rangeTableList)
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
+
+		if (rangeTableEntry->rtekind == RTE_RELATION)
+		{
+			Oid relationId = rangeTableEntry->relid;
+			bool distributedTable = IsDistributedTable(relationId);
+
+			if (!distributedTable)
+			{
+				return false;
+			}
+
+			if (PartitionMethod(relationId) != DISTRIBUTE_BY_NONE)
+			{
+				return false;
+			}
+		}
+		else
+		{
+			return false;
+		}
+	}
```

Comment:
I wonder if the following style (which is the preferred style in some other companies) is still unacceptable in Citus:

```C
        if (rangeTableEntry->rtekind != RTE_RELATION ||
		    !IsDistributedTable(rangeTableEntry->relid) ||
			PartitionMethod(rangeTableEntry->relid) != DISTRIBUTE_BY_NONE)
		{
			return false;
		}
```

Reply:
@pykello You wrote the coding conventions document so you can guide us towards new conventions :smiley_cat: